### PR TITLE
Hosting deploy single file script

### PR DESCRIPTION
**Pull request by bkendall**
### Description

This is an example script for deploying to Hosting using the REST API without relying on `firebase-tools`. Instead of piggybacking off `firebase-tools`'s auth, this uses `google-auth-library` and ADC (or any other way the library can auth) to get the user set up and running.

### CODE REVIEW

File: scripts/examples/hosting/update-single-file/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+gcloud auth application-default login
+# It may be required to set a quota project for the credentials - used to account for the API usage.
+gcloud auth application-default set-quota-project <project-id>
+```
+
+Alternatively, if you (want to) use a service account and set `GOOGLE_APPLICATION_CREDENTIALS` instead of using `gcloud`, that works well too. See Google Cloud's [getting started with authentication](https://cloud.google.com/docs/authentication/getting-started) for more infromation on how to set one up.
+
+### Run the Script
+
+In your `public` directory (with your Hosting files):
```
Comment (puf) [inline scripts/examples/hosting/update-single-file/README.md:37]: Maybe: "In the directory that you specified as `public` in your Firebase Hosting configuration:"
Trying to steer away from the word "files" here, as I misunderstood initially and only later saw that you specify the files to add/update later.

File: scripts/examples/hosting/update-single-file/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+```bash
+cd my-app/public/
+update-single-file --project <project-id> [--site <site-id>] <files...>
+```
+
+Running the script _from your `public` directory is important_. For example, if you want to update `/team/about.html` in your site you would:
```
Comment (puf) [inline scripts/examples/hosting/update-single-file/README.md:44]: I'd skip "Running the script _from your `public` directory is important_.". You said it above already.

Of course, I might be wrong - in which we can add it back later and I'll bow my head in shame.

**Review by puf — COMMENTED** [review]
Left two nits on the README. 
LGTM

File: scripts/examples/hosting/update-single-file/README.md (hunk 2 above)
Comment (bkendall) [reply scripts/examples/hosting/update-single-file/README.md:44]: Deal!

File: scripts/examples/hosting/update-single-file/README.md (hunk 1 above)
Comment (bkendall) [reply scripts/examples/hosting/update-single-file/README.md:37]: Good call. Thanks

File: scripts/examples/hosting/update-single-file/README.md (hunk 2 above)
Comment (bkendall) [reply scripts/examples/hosting/update-single-file/README.md:44]: Thanks

File: scripts/examples/hosting/update-single-file/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  const SITE_ID = argv.site || PROJECT_ID;
+  const files = Array.from(argv._);
+  console.log(`Deploying files:`);
+  for (const f of files) {
+    console.log(`- ${f}`);
+  }
+
+  const filesByHash: Record<string, string> = {};
+  for (const file of files) {
+    const hasher = crypto.createHash("sha256");
```
Comment (mbleigh) [inline scripts/examples/hosting/update-single-file/src/index.ts:51]: Maybe abstract the "gzip and hash a single file" part into a function? Would be helpful to have it separated for exemplar reasons I think.

File: scripts/examples/hosting/update-single-file/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  let done = false;
+  let newVersion = "";
+  while (!done) {
+    const opRes = await client.request<{ done: boolean; response: { name: string } }>({
+      url: `${HOSTING_URL}/${operationName}`,
+    });
+    debug("%d %j", opRes.status, opRes.data);
+    done = !!opRes.data.done;
+    newVersion = opRes.data.response?.name;
+    await new Promise((resolve) => setTimeout(resolve, 500));
```
Comment (mbleigh) [inline scripts/examples/hosting/update-single-file/src/index.ts:117]: I'd probably have this be at least 1s

**Review by mbleigh — APPROVED** [review]
In general you might want to factor out some named methods for readability, especially if this is an "example script", and maybe toss in some more comments to show what's happening. Nothing stood out as a problem with the behavior tho.

File: scripts/examples/hosting/update-single-file/src/index.ts (hunk 4 above)
Comment (bkendall) [reply scripts/examples/hosting/update-single-file/src/index.ts]: Sure.

File: scripts/examples/hosting/update-single-file/src/index.ts (hunk 3 above)
Comment (bkendall) [reply scripts/examples/hosting/update-single-file/src/index.ts]: Done.

**Pull request merged**